text/0224-experimental-semantic-conventions.md

```diff
@@ -0,0 +1,117 @@
+
+## Definitions
+
+**Attribute Dictionary**: The attribute dictionary contains all semantic convention attributes, including their names and value definitions.
+
+**Registered Attribute**: A registered attribute is any attribute which is in the attribute dictionary.
+
+## Proposal
+
+A semantic convention attribute dictionary will be established which is maintained in a separate file or directory from the semantic conventions, seeded with all existing semantic convention attributes.
+All semantic conventions MUST refer to the attribute dictionary for all attributes used by that semantic convention.
+Semantic conventions MUST NOT establish an attribute not in the attribute dictionary without explicit approval of the TC with the reasoning documented directly in that convention.
+A procedure will be established by the technical committee (TC), or an entity designated by the TC, for registration of new semantic convention attributes.
+Any attributes not intended strictly for private use SHOULD be registered in the attribute dictionary.
+
+Instrumentation authors SHOULD assume that any attribute they create may become standardized, public, commonly deployed, or usable across multiple OpenTelemetry clients, distributions, or receivers.
+They SHOULD choose attribute keys that are descriptive, relevant, and that they have reason to believe are not currently in use or already proposed.
+Before creating a new attribute, the attribute author SHOULD review the attribute dictionary to see if an attribute already exists which fits the intended use case.
+If such a registered attribute exists, then it SHOULD be preferred over the creation of new attributes unless there is sufficient reason to create a new attribute.
+Any new attributes SHOULD be registered in the attribute dictionary before an instrumentation using the attribute is released.
+If an instrumentation uses an attribute not included in the semantic conventions it is recommended that the attribute not be collected by default unless it is fundamental to the instrumentation.
+For example, an instrumentation which instruments a component not yet addressed by semantic conventions.
+
+An entry MUST NOT be refused without credible reason to believe that registration would be harmful.
+Note that a disagreement about technical details are not considered grounds to refuse listing in the attribute dictionary.
+Some examples of reasonable grounds for refusal: frivolous use of the dictionary such as registering nonsense words or inside jokes, a credible security or privacy risk, conflict with another attribute or proposal for a new attribute, or that the proposal is sufficiently lacking in purpose, or misleading about its purpose, that it can be held to be a waste of time and effort.
+This list is non-exhaustive and final decision for refusal is at the discretion of the TC.
+
+Once an attribute is registered it MUST NOT receive any breaking changes.
```

Comment:
We need some time to experiment with attributes before they stabilize and we already have `experimental` and `stable` maturity levels defined in the build tools. 
Can we define them explicitly here and allow experimental attributes to get breaking changes?

Reply:
I really don't want that. I would much prefer a new attribute to be registered with a new name if a breaking change really needs to be made. Once an attribute is "in the wild" it should be treated as stable. If we register experimental attributes we're going to end up in the same situation we are now with the semconv where they sit for years in experimental then one day get broken and users are broken.

What if instead I update the language such that an attribute in early experimentation phase should _not_ be registered until the author is confident the format will not have a breaking change?

Reply:
see my other comment below - we renamed attributes a lot in the past (and made other breaking changes to them), there is no reason to believe we would not continue this.

> What if instead I update the language such that an attribute in early experimentation phase should not be registered until the author is confident the format will not have a breaking change?

then we need to allow experimental semconv to use their own, non-registered attributes until they are happy with them.

Reply:
> then we need to allow experimental semconv to use their own, non-registered attributes.

My concern with that is that we'll end up with a sort of chaotic state where parts of attributes are registered and others remain with their definition in the semconv for a long time. So the dictionary is then not really a true dictionary because it will miss many attributes.

Reply:
>My concern with that is that we'll end up with a sort of chaotic state where parts of attributes are registered and others remain with their definition in the semconv for a long time. 

That is exactly what the registry is to prevent and why the barrier to entry is so low.

>then we need to allow experimental semconv to use their own, non-registered attributes.

I'd like to prevent that as much as we can.

> see my other comment below - we renamed attributes a lot in the past (and made other breaking changes to them), there is no reason to believe we would not continue this.

If an attribute needs to be renamed, a new attribute can simply be registered and the old one deprecated. The new attribute is then used in the semantic convention. I'm not sure what the concern is.